server-spi-private/src/main/java/org/keycloak/userprofile/UserProfileContext.java

```diff
@@ -1,31 +1,43 @@
 /*
- * Copyright 2020 Red Hat, Inc. and/or its affiliates
- * and other contributors as indicated by the @author tags.
  *
- * Licensed under the Apache License, Version 2.0 (the "License");
- * you may not use this file except in compliance with the License.
- * You may obtain a copy of the License at
+ *  * Copyright 2021  Red Hat, Inc. and/or its affiliates
+ *  * and other contributors as indicated by the @author tags.
+ *  *
+ *  * Licensed under the Apache License, Version 2.0 (the "License");
+ *  * you may not use this file except in compliance with the License.
+ *  * You may obtain a copy of the License at
+ *  *
+ *  * http://www.apache.org/licenses/LICENSE-2.0
+ *  *
+ *  * Unless required by applicable law or agreed to in writing, software
+ *  * distributed under the License is distributed on an "AS IS" BASIS,
+ *  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  * See the License for the specific language governing permissions and
+ *  * limitations under the License.
  *
- * http://www.apache.org/licenses/LICENSE-2.0
- *
- * Unless required by applicable law or agreed to in writing, software
- * distributed under the License is distributed on an "AS IS" BASIS,
- * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
- * See the License for the specific language governing permissions and
- * limitations under the License.
  */
 
 package org.keycloak.userprofile;
 
-import org.keycloak.userprofile.validation.UserProfileValidationResult;
-import org.keycloak.userprofile.validation.UserUpdateEvent;
-
 /**
+ * <p>This interface represents the different contexts from where user profiles are managed. The core contexts are already
+ * available here representing the different parts in Keycloak where user profiles are managed.
+ *
+ * <p>The context is crucial to drive the conditions that should be respected when managing user profiles. It might be possible
+ * to include in the future metadata about contexts. As well as support custom contexts.
+ *
  * @author <a href="mailto:markus.till@bosch.io">Markus Till</a>
  */
 public interface UserProfileContext {
 
-    UserUpdateEvent getUpdateEvent();
-    UserProfile getCurrentProfile();
-    UserProfileValidationResult validate();
+    UserProfileContext UPDATE_PROFILE = create();
```

Comment:
We should maybe extend these contexts so they clearly state if they are about existing user update, new user creation or existing user check. This way code in provider can be much clearer for these important business cases. Now you are guessing it at some places by checks like 'user==null' combined with these detailed contexts.

Reply:
Not sure. It is not a guess but how it works. 

If you create a profile with a user, that means you have a profile for an existing user. Otherwise, you are either validating a profile or creating a new user.

IMO, having a fine-grained definition of contexts just makes things more complicated.